test/unit/key_def.cc

```diff
@@ -457,16 +460,1307 @@ test_tuple_validate_key_parts_raw(void)
+	struct key_def *key_def = test_key_def_new(
+		"[{%s%u%s%s%s%b}]", "field", 1, "type", "number",
+		"is_nullable", is_nullable_and_has_optional_parts);
+
+	/* Update has_optional_parts if the last parts can't be nil. */
+	size_t min_field_count = tuple_format_min_field_count(&key_def, 1,
+							      NULL, 0);
+	key_def_update_optionality(key_def, min_field_count);
+
+	const char *funcname = (const char *)xstrdup(tt_sprintf(
+		"tuple_compare_with_key_slowpath<%s, key_def.part_count = 1>",
+		is_nullable_and_has_optional_parts ?
+		"true, true" : "false, false"));
+
+	std::vector<struct tuple *> tuples_eq = {
+		/* Regular case. */
+		test_tuple_new("[%u%u]", 0, 0),
+		test_tuple_new("[%u%u]", 0, 0),
+
+		/* The first field is not indexed. */
+		test_tuple_new("[%u%u]", 1, 0),
+		test_tuple_new("[%u%u]", 0, 0),
+	};
+
+	if (is_nullable_and_has_optional_parts) {
+		/* NILs and unexisting parts. */
+		tuples_eq.push_back(test_tuple_new("[%uNIL]", 0));
+		tuples_eq.push_back(test_tuple_new("[%uNIL]", 0));
+
+		tuples_eq.push_back(test_tuple_new("[%u]", 0));
+		tuples_eq.push_back(test_tuple_new("[%uNIL]", 0));
+
+		tuples_eq.push_back(test_tuple_new("[%u]", 0));
+		tuples_eq.push_back(test_tuple_new("[%u]", 0));
+	}
+
+	std::vector<struct tuple *> tuples_gt = {
+		/* Regular cases. */
+		test_tuple_new("[%u%u]", 0, 1),
+		test_tuple_new("[%u%u]", 0, 0),
+
+		/* The first field is not indexed. */
+		test_tuple_new("[%u%u]", 0, 1),
+		test_tuple_new("[%u%u]", 1, 0),
+	};
+
+	if (is_nullable_and_has_optional_parts) {
+		/* NILs and unexisting parts. */
+		tuples_gt.push_back(test_tuple_new("[%u%u]", 0, 0));
+		tuples_gt.push_back(test_tuple_new("[%uNIL]", 0));
+
+		tuples_gt.push_back(test_tuple_new("[%u%u]", 0, 0));
+		tuples_gt.push_back(test_tuple_new("[%u]", 0));
+	}
+
+	assert(tuples_eq.size() % 2 == 0);
+	for (size_t i = 0; i < tuples_eq.size(); i += 2) {
+		struct tuple *a = tuples_eq[i];
+		struct tuple *b = tuples_eq[i + 1];
+		test_tuple_compare_with_key(a, b, key_def, 0, funcname);
+		test_tuple_compare_with_key(b, a, key_def, 0, funcname);
+	}
+
+	assert(tuples_gt.size() % 2 == 0);
+	for (size_t i = 0; i < tuples_gt.size(); i += 2) {
+		struct tuple *a = tuples_gt[i];
+		struct tuple *b = tuples_gt[i + 1];
+		test_tuple_compare_with_key(a, b, key_def, 1, funcname);
+		test_tuple_compare_with_key(b, a, key_def, -1, funcname);
+	}
+
+	for (size_t i = 0; i < tuples_eq.size(); i++) {
+		tuple_delete(tuples_eq[i]);
+	}
+
+	for (size_t i = 0; i < tuples_gt.size(); i++) {
+		tuple_delete(tuples_gt[i]);
+	}
+
+	key_def_delete(key_def);
+
+	free((void *)funcname);
+
+	footer();
+	check_plan();
+}
+
+static void
+test_tuple_compare_with_key_slowpath(bool is_nullable, bool has_optional_parts)
+{
+	size_t p = 14 + (is_nullable ? 10 : 0) + (has_optional_parts ? 40 : 0);
+	plan(p);
+	header();
+
+	/* has_optional_parts is only valid if is_nullable. */
+	assert(!has_optional_parts || is_nullable);
```

Comment:
Nit: We usually use `fail_if` and `fail_unless` functions to check invariants in unit tests.

Reply:
Replaced the asserts with `fail_unless`.